Code/GraphMol/SubstanceGroup.cpp

```diff
@@ -197,6 +296,123 @@ unsigned int addSubstanceGroup(ROMol &mol, SubstanceGroup sgroup) {
   return id;
 }
 
+namespace {
+bool includesBond(SubstanceGroup &sg, unsigned int idx) {
+  return sg.includesBond(idx);
+}
+bool includesAtom(SubstanceGroup &sg, unsigned int idx) {
+  return sg.includesAtom(idx);
+}
+void removedBond(SubstanceGroup &sg, unsigned int idx) {
+  sg.adjustToRemovedBond(idx);
+}
+void removedAtom(SubstanceGroup &sg, unsigned int idx) {
+  sg.adjustToRemovedAtom(idx);
+}
+
+bool removedParentInHierarchy(
+    unsigned int idx, const std::vector<SubstanceGroup> &sgs,
+    const boost::dynamic_bitset<> &toRemove,
+    const std::map<unsigned int, unsigned int> &indexLookup) {
+  PRECONDITION(idx < sgs.size(), "cannot find SubstanceGroup");
+  if (toRemove[idx]) return true;
+
+  unsigned int parent;
+  if (sgs[idx].getPropIfPresent("PARENT", parent)) {
+    auto piter = indexLookup.find(parent);
+    if (piter != indexLookup.end()) {
+      return removedParentInHierarchy(piter->second, sgs, toRemove,
+                                      indexLookup);
+    }
+  }
+  return false;
+}
+
+template <bool INCLUDES_METHOD(SubstanceGroup &, unsigned int),
+          void ADJUST_METHOD(SubstanceGroup &, unsigned int)>
+void removeSubstanceGroupsReferencing(RWMol &mol, unsigned int idx) {
+  auto &sgs = getSubstanceGroups(mol);
+  if (!sgs.empty()) {
+    // first collect the ones that should be removed
+    boost::dynamic_bitset<> toRemove(sgs.size());
+    unsigned int nRemoved = 0;
+    bool parentsPresent = false;
+    for (unsigned int i = 0; i < sgs.size(); ++i) {
+      if (!parentsPresent && sgs[i].hasProp("PARENT")) {
+        parentsPresent = true;
+      }
+      if (INCLUDES_METHOD(sgs[i], idx)) {
+        toRemove.set(i);
+        ++nRemoved;
+      }
+    }
+
```

Comment:
I'd split off `toRemove` in the next `if` to a separate check here:
```suggestion
    if (toRemove.none()) {
        return;
    }
```
to exit early if no SubstanceGroups were to be removed.

Also, I think we don't need `nRemoved`, since we can use the dynamic_bitset's methods (`none()`, `count()`, etc).

Reply:
yeah, I thought about returning early here too, but that doesn't work since we still need to update the atom/bond indices in the SubstanceGroups that we keep

Reply:
Oh, right.